torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -478,29 +484,51 @@ void ProcessGroupNCCL::ncclCommWatchdogInternal() {
-              // other ranks might have failed/timed out and didn't enter
-              // `ncclCommInitRank`. As a result, when there is a failure on
-              // a communicator the application receives an exception and its
-              // their responsibility to destroy the process group and recreate
-              // it to recover from errors.
-              abortedCommIds.emplace(
-                  buildNcclUniqueIdStr(ncclComm->getNcclId()));
-            }
+          LOG(INFO) << "Aborting communicators that received errors";
+          // We should not abort the communicators if we are performing a
+          // non-blocking wait(). The reason for this is that if we abort the
+          // nccl communicator, wait() might not throw exceptions and
+          // subsequent operations might run on garbage results.
+          // The current model is that when we call wait(), subsequent
+          // operations only run after this work is done or we hang forever
+          // waiting for the operation to complete.
+          for (const auto& ncclComm : ncclComms) {
+            ncclComm->ncclCommAbort();
+            // Note that we don't remove the aborted communicators from the
+            // cache. The reason is that if we do remove the communicator
+            // from the cache, it is possible that a new collective operation
+            // calls `ncclCommInitRank` to create a new communicator whereas
+            // other ranks might have failed/timed out and didn't enter
+            // `ncclCommInitRank`. As a result, when there is a failure on
+            // a communicator the application receives an exception and its
+            // their responsibility to destroy the process group and recreate
+            // it to recover from errors.
+            abortedCommIds.emplace(buildNcclUniqueIdStr(ncclComm->getNcclId()));
+          }
+        }
+      }
+    }
+
+    {
+      std::unique_lock<std::mutex> lock(workListMutex_);
+      for (auto& work : workList_) {
+        work->checkAndSetException();
+        // Aborting NCCL Communicators due to errors is already handled above.
+        if (work->exception()) {
+          continue;
+        }
+
+        // Check for Timeouts in the WorkNCCL Operations, and abort all
+        // communicators accordingly.
+        auto currentTimepoint = std::chrono::steady_clock::now();
+        if (std::chrono::duration_cast<std::chrono::milliseconds>(
+                currentTimepoint - work->workStartTime_) > work->opTimeout_) {
+          std::exception_ptr exception_ptr = std::make_exception_ptr(
+              std::runtime_error("NCCL Operation Timed Out"));
+          work->setException(exception_ptr);
+          for (const auto& ncclComm : work->ncclComms_) {
+            ncclComm->ncclCommAbort();
+            abortedCommIds.emplace(buildNcclUniqueIdStr(ncclComm->getNcclId()));
           }
         }
       }
```

Comment:
This function is becoming pretty large with a multiple complex blocks, can we move each block out into separate helper functions for more clarity?